### PR TITLE
[deckhouse] remove module external-module-manager

### LINKED ISSUE #5545: [external-module-manager] Rebranding
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

The name `external` influences the attitude of users towards modules. They start thinking that `external` modules are not actual modules but leftovers. The idea behind the external module manager was to abstain from writing any more embedded modules and only use a new delivery mechanism.

### Proposed Solution

Rename `external-modules` to `modules`.

- [x] Rename CRD: ExternalModuleSource -> ModuleSource, ExternalModuleRelease -> ModuleRelease.
- [ ] Remove the module `external-module-manager`.
  - [ ] Move the hook `cleanup_releases` to the ModuleReleaseController
- [ ] Rename mounts: `/var/lib/deckhouse/external-modules/` -> `/var/lib/deckhouse/modules/`
- [x] Rename in addon-operator

### Additional Information

For migration of resources:
- Add new CRD: `ModuleSource` and `ModuleRelease`
- Add an onstartup hook to copy `ExternalModule*` custom resources content to `Module*`
- Add a validating webhook to block create requests of ExternalModuleSource and ExternalModuleRelease, suggest creating `ModuleSource` and `ModuleRelease` instead

It will break deploying from Git, but regards external modules not being widely used, it seems a good trade-off for the implementation speed

### PR DESCRIPTION
**Pull request by miklezzzz**
## Description
The following changes are introduced:
1) `external-module-manager` is removed and it's cleanup hook is re-implemented in the `ModuleSource` controller;
It's assumed that the functionality of the deleted module is always available (in any bundle for any cluster). 
The deckhouse deployment template is adjusted to support new variables and directories for downloadable modules. To make the migration from `external-modules` directory to `downloaded` smoother, an env `GetDownloadedModulesDir` helper is introduced that returns the value of `DOWNLOADED_MODULES_DIR` env if it exists and fall backs to `EXTERNAL_MODULES_DIR` in the opposite case. The init container's script is updated to move the content of `external-modules` directory to `downloaded` directory;
2) `dhctl` code base is updated according to new naming;
3) `ensure_crd` package is updated to support deletion of outdated CRD's (like `ExternalModuleRelease`/`ExternalModuleSource`. CRDs are deleted from the cluster only if there are no objects of the associated kinds left in the cluster;
4) `ModuleSource` controller's tests are fixed (didn't check the whole golden file, but only first manifest);
5) Relevant tests are introduced to check new functionality;
6) Deprecated global hooks `module_migration` are deleted.

## Why do we need it, and what problem does it solve?
Closes #5545 
This pr deprecates the `external module` notion by removing `external-module-manager` and re-implementing part of its functionality (clean up outdated `ModuleReleases`) in the `ModuleSource` controller.

## What is the expected result?
There is no `external-module-manager` in the cluster.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: chore
summary: Remove external-module-manager module and deprecate `external` notion.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by miklezzzz** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 9973954965 9973954965-1-con-1-27 /sys/deckhouse-oss/install:pr9051 redos@178.154.201.211

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/openstack'...

**Comment by miklezzzz** [comment]
/e2e/abort openstack;Standard;containerd;1.27 10004415371 10004415371-1-con-1-27 /sys/deckhouse-oss/install:pr9051 redos@46.161.52.49

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by miklezzzz** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 10005569712 10005569712-1-con-1-27 /sys/deckhouse-oss/install:pr9051 redos@89.169.134.218

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/gcp'...

**Review by z9r5 — APPROVED** [review]

File: dhctl/pkg/kubernetes/actions/manifests/manifests.go [hunk 1]
```diff
@@ -282,25 +290,25 @@ func DeckhouseDeployment(params DeckhouseDeploymentParams) *appsv1.Deployment {
...
 	}
 
 	deckhouseInitContainer := apiv1.Container{
-		Name:            "init-external-modules",
+		Name:            "init-downloaded-modules",
 		Image:           initContainerImage,
 		ImagePullPolicy: apiv1.PullAlways,
 		Command: []string{
-			"sh", "-c", "mkdir -p /deckhouse/external-modules/modules && chown -hR 64535 /deckhouse/external-modules /deckhouse/external-modules/modules && chmod 0700 /deckhouse/external-modules /deckhouse/external-modules/modules",
+			"sh", "-c", `if [ -d "/host/deckhouse/external-modules" ] && [ -n "$(ls -A "/host/deckhouse/external-modules")" ]; then cp -r /host/deckhouse/external-modules/* /host/deckhouse/downloaded/ && rm -rf /host/deckhouse/external-modules; fi && mkdir -p /host/deckhouse/downloaded/modules && chown -hR 64535 /host/deckhouse/downloaded /host/deckhouse/downloaded/modules && chmod 0700 /host/deckhouse/downloaded /host/deckhouse/downloaded/modules`,
```
Comment (yalosev) [inline dhctl/pkg/kubernetes/actions/manifests/manifests.go:305]: `/host`? What is it?

File: deckhouse-controller/pkg/controller/module-controllers/source/controller.go [hunk 2]
```diff
@@ -216,6 +219,11 @@ func (c *moduleSourceReconciler) createOrUpdateReconcile(ctx context.Context, ms
 	// save checksums
 	c.saveSourceChecksums(ms.Name, modulesChecksums)
 
+	err = c.cleanUpModuleReleases(ctx, ms)
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/module-controllers/source/controller.go:222]: move to module-release controller

File: deckhouse-controller/pkg/controller/module-controllers/source/controller.go [hunk 3]
```diff
@@ -419,31 +427,98 @@ func (c *moduleSourceReconciler) updateModuleSourceStatus(ctx context.Context, m
...
+		name    string
+		version *semver.Version
+	}
+
+	outdatedReleases := make(map[string][]outdatedRelease, 0)
+
+	// get all outdated releases by module names
+	for _, rl := range moduleReleasesFromSource.Items {
+		if metav1.IsControlledBy(&rl, ms) {
+			if rl.Status.Phase == v1alpha1.PhaseSuperseded || rl.Status.Phase == v1alpha1.PhaseSuspended {
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/module-controllers/source/controller.go:449]: + phase.Skipped

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/gcp'...

**Comment by miklezzzz** [comment]
```
{"component":"ModuleReleaseController","level":"info","msg":"Applying release echo-v0.4.65","time":"2024-08-15T15:28:56Z"}
{"level":"info","msg":"Validating module \"echo\" from \"/deckhouse/downloaded/echo/v0.4.65\"","time":"2024-08-15T15:28:56Z"}
{"level":"warning","msg":"schema (echo) for 'values' values is not found","time":"2024-08-15T15:28:56Z"}
{"level":"warning","msg":"schema (echo) for 'config' values is not found","time":"2024-08-15T15:28:56Z"}
{"component":"ModuleDocumentation","level":"info","msg":"Updating documentation for echo module","time":"2024-08-15T15:29:00Z"}
{"component":"ModuleReleaseController","level":"info","msg":"cleaned up outdated release \"echo-v0.4.61\" of \"echo\" module","time":"2024-08-15T15:29:00Z"}
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by yalosev — APPROVED** [review]

**Pull request merged**